plugins/field-dependent-dropdown/src/field_dependent_dropdown.ts

```diff
@@ -0,0 +1,274 @@
+      // If derivedOptions has been initialized, use that.
+      if (dependencyData.derivedOptions) {
+        return dependencyData.derivedOptions;
+      }
+
+      // Fall back on the options corresponding to the parent field's current
+      // value (which is fine when initializing but may be out of date when
+      // making changes since the parent field's validator function triggers
+      // this function before the parent field's value is updated).
+
+      // If the parent field exists, and its value is a key in the provided
+      // option mapping, use the corresponding options.
+      if (dependencyData.parentField) {
+        const options = optionMapping[dependencyData.parentField.getValue()];
+        if (options) {
+          return options;
+        }
+      }
+
+      // Fall back on basic default options.
+      return [['', '']];
+    };
+
+    super(menuGenerator, validator, config);
+    this.parentName = parentName;
+    this.optionMapping = optionMapping;
+    this.dependencyData = dependencyData;
+  }
+
+  /**
+   * Constructs a FieldDependentDropdown from a JSON arg object.
+   * @param options A JSON object providing "parentName" and "optionMapping".
+   * @returns The new field instance.
+   */
+  static fromJson(
+      options: FieldDependentDropdownFromJsonConfig): FieldDependentDropdown {
+    return new FieldDependentDropdown(
+        options['parentName'],
+        options['optionMapping'],
+        undefined,
+        options);
+  }
+
+  /**
+   * Attach this field to a block.
+   *
+   * @param block The block containing this field.
+   */
+  setSourceBlock(block: Blockly.Block) {
+    super.setSourceBlock(block);
+
+    const parentField: Blockly.Field<string> | null =
+        block.getField(this.parentName);
+
+    if (!parentField) {
+      throw new Error('Could not find a parent field with the name ' +
+          this.parentName + ' for the dependent dropdown.');
+    }
+
+    this.dependencyData.parentField = parentField;
+
+    const oldValidator = parentField.getValidator();
+
+    // A validator function for the parent field that has the side effect of
+    // updating the options of this child dropdown field based on the new value
+    // of the parent field whenever it changes. The validator function is a good
+    // place to do this because it is called immediately while deserializing
+    // workspaces before the following fields are deserialized, so when the
+    // child value is deserialized the appropriate options will already be
+    // available. If the parent already had a validator function, it will be
+    // composed with this one and the new value returned from it will be the
+    // basis for determining the new available options.
+    parentField.setValidator((newValue) => {
+      if (oldValidator) {
+        const validatedValue = oldValidator(newValue);
+        // If a validator returns undefined, that means no change. Otherwise,
```

Comment:
This is not quite right. The validator can return one of three types of values:
- `null` means the value was invalid. If the original parent validator returns null, we also want to return null from this validator, and not update the options, because the value is not actually going to change. The parent field's `doValueInvalid_` method will be called, but that's outside the scope of the child field. From the child field's perspective, the value isn't changing and we shouldn't update anything.
- `undefined` means set the `newValue` exactly it was proposed, e.g. `newValue = newValue`
- any other value means set the `newValue` to the value returned from the validator, e.g. `newValue = validatedValue`.

So you need to handle all three cases here, by adding

```
if (validatedValue === null) return null;
```